### PR TITLE
docs: add annotation for `phpstan`

### PR DESCRIPTION
**Pull request by datamweb**
I try in this PR, To fix the errors identified by `phpstan` at **level 9**.

This is my first experience, anyone please help.

 :man_facepalming: all errors : 154

### CODE REVIEW

**Comment by datamweb** [comment]
**Before** 
```console 
 ------ ---------------------------------------------------------------------------------------------------------------
  Line   ShieldOAuthConfig.php
 ------ ---------------------------------------------------------------------------------------------------------------
  20     Property Datamweb\ShieldOAuth\Config\ShieldOAuthConfig::$oauthConfigs type has no value type specified in
         iterable type array.
         💡 See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
  56     Property Datamweb\ShieldOAuth\Config\ShieldOAuthConfig::$usersColumnsName type has no value type specified in
         iterable type array.
         💡 See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
 ------ --------------------------------------------------------------------------------------------------------------- 
```
**After** 
```console 
 1/1 [============================] 100%

 [OK] No errors
```

**Comment by datamweb** [comment]
**Before:**
```console
> phpstan analyze src/Config/Registrar.php

 1/1 [============================] 100%

 ------ ----------------------------------------------------------------------------------------------------------------------
  Line   Registrar.php
 ------ ----------------------------------------------------------------------------------------------------------------------
  30     Method Datamweb\ShieldOAuth\Config\Registrar::View() return type has no value type specified in iterable type array.
         💡 See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
 ------ ----------------------------------------------------------------------------------------------------------------------

 [ERROR] Found 1 error
```
**After:** 
```console
> phpstan analyze src/Config/Registrar.php

 1/1 [============================] 100%

 [OK] No errors
```

File: src/Config/Registrar.php [hunk 1]
```diff
@@ -18,9 +18,16 @@
...
 {
     /**
+     * --------------------------------------------------------------------------
+     * Shield OAuth Decorator
+     * --------------------------------------------------------------------------
      * Register the ShieldOAuth decorator.
+     *
+     * @see https://codeigniter.com/user_guide/outgoing/view_decorators.html
+     *
+     * @return array<string, array<int, string>>
```
Comment (MGatner) [inline src/Config/Registrar.php]: PHPStan has lists now too! This is slightly more specific (both are fine though):
```suggestion
     * @return array<string, list<string>>
```

Comment (datamweb) [reply src/Config/Registrar.php]: Thanks 👍 
Anybody see [List](https://phpstan.org/writing-php-code/phpdoc-types#lists) for more info.

**Comment by datamweb** [comment]
**Before:**
```console
P:\WorkApp\cv472\vendor\datamweb\shield-oauth>composer analyze src\Config\Services.php
 1/1 [============================] 100%

 ------ --------------------------------------------------------------------------------------------------
  Line   Services.php
 ------ --------------------------------------------------------------------------------------------------
  21     Method Datamweb\ShieldOAuth\Config\Services::ShieldOAuth() has no return type specified.
  21     Method Datamweb\ShieldOAuth\Config\Services::ShieldOAuth() has parameter $getShared with no type
         specified.
  27     Class Datamweb\ShieldOAuth\Libraries\Basic\ShieldOAuth does not have a constructor and must be
         instantiated without any parameters.
 ------ --------------------------------------------------------------------------------------------------
```

**After:**
```console
> phpstan analyze src\Config\Services.php

Note: Using configuration file P:\WorkApp\cv472\vendor\datamweb\shield-oauth\phpstan.neon.dist.
 1/1 [============================] 100%

 [OK] No errors

```

**Comment by datamweb** [comment]
**Before:**
 ```console
------ ---------------------------------------------------
  Line   Routes.php
 ------ ---------------------------------------------------
  14     Variable $routes might not be defined.
  15     Cannot call method allOAuth() on mixed.
  18     Cannot access property $call_back_route on mixed.
 ------ ---------------------------------------------------


 [ERROR] Found 3 errors
```
**After:**
```console
> phpstan analyze --xdebug src/Config/Routes.php
Note: Using configuration file P:\WorkApp\cv472\vendor\datamweb\shield-oauth\phpstan.neon.dist.
 1/1 [============================] 100%


 [OK] No errors
```
I got help from **kenjis** & **paulbalandan** , so thanks to them.

**Comment by datamweb** [comment]
**Before:**
```console
 ------ ----------------------------------------------------
  Line   ShieldOAuthModel.php
 ------ ----------------------------------------------------
  28     Cannot access property $usersColumnsName on mixed.
  29     Cannot access property $usersColumnsName on mixed.
  30     Cannot access property $usersColumnsName on mixed.
 ------ ----------------------------------------------------
```
**After:**
```console
 0/1 [>---------------------------]   0%

 [OK] No errors
```

**Comment by datamweb** [comment]
**Before:**
```console
 ------ ----------------------------------------------------------------------------------------------------
  Line   2022-10-20-182737_ShieldOAuth.php
 ------ ----------------------------------------------------------------------------------------------------
  24     Access to an undefined property Datamweb\ShieldOAuth\Database\Migrations\ShieldOAuth::$first_name.
  24     Cannot access property $usersColumnsName on mixed.
  25     Access to an undefined property Datamweb\ShieldOAuth\Database\Migrations\ShieldOAuth::$last_name.
  25     Cannot access property $usersColumnsName on mixed.
  26     Access to an undefined property Datamweb\ShieldOAuth\Database\Migrations\ShieldOAuth::$avatar.
  26     Cannot access property $usersColumnsName on mixed.
  32     Access to an undefined property Datamweb\ShieldOAuth\Database\Migrations\ShieldOAuth::$first_name.
  37     Access to an undefined property Datamweb\ShieldOAuth\Database\Migrations\ShieldOAuth::$last_name.
  42     Access to an undefined property Datamweb\ShieldOAuth\Database\Migrations\ShieldOAuth::$avatar.
  56     Access to an undefined property Datamweb\ShieldOAuth\Database\Migrations\ShieldOAuth::$first_name.
  57     Access to an undefined property Datamweb\ShieldOAuth\Database\Migrations\ShieldOAuth::$last_name.
  58     Access to an undefined property Datamweb\ShieldOAuth\Database\Migrations\ShieldOAuth::$avatar.
 ------ ----------------------------------------------------------------------------------------------------

 [ERROR] Found 12 errors
```
**After:**
```console
 1/1 [============================] 100%

 [OK] No errors
```

**Comment by datamweb** [comment]
**Before:**
```console
> phpstan analyze --xdebug src\Libraries\Basic\ControllersInterface.php
 1/1 [============================] 100%

 ------ -----------------------------------------------------------------------------------------------------------------
  Line   ControllersInterface.php
 ------ -----------------------------------------------------------------------------------------------------------------
  19     Method Datamweb\ShieldOAuth\Libraries\Basic\ControllersInterface::redirectOAuth() has no return type specified.
  21     Method Datamweb\ShieldOAuth\Libraries\Basic\ControllersInterface::callBack() has no return type specified.
 ------ -----------------------------------------------------------------------------------------------------------------

 [ERROR] Found 2 errors
```
**After:**
```console
> phpstan analyze --xdebug src\Libraries\Basic\ControllersInterface.php
 1/1 [============================] 100%

 [OK] No errors

```

**Comment by datamweb** [comment]
**Before:**
```console
 ------ ---------------------------------------------------------------------------------------------------------------------------------------
  Line   ShieldOAuth.php
 ------ ---------------------------------------------------------------------------------------------------------------------------------------
  31     Call to an undefined static method CodeIgniter\Config\Factories::loadOAuth().
  48     Cannot access property $psr4 on mixed.
  72     Method Datamweb\ShieldOAuth\Libraries\Basic\ShieldOAuth::otherOAuth() return type has no value type specified in iterable type array.
         💡 See: https://phpstan.org/blog/solving-phpstan-no-value-type-specified-in-iterable-type
 ------ ---------------------------------------------------------------------------------------------------------------------------------------


 [ERROR] Found 3 errors
```
**After:**
```console
 1/1 [============================] 100%


 [OK] No errors
```

**Note:**

https://github.com/datamweb/shield-oauth/blob/0be15d39d1ea40599c510a85d3b3283228b57a36/src/Libraries/Basic/ShieldOAuth.php#L31

I tried my best not to use `@phpstan-ignore-line`, but method `loadOAuth()` doesn't really exist. So there seems to be no other way than using `@phpstan-ignore-line` to have it rejected by `PHPStan`.
However, I am not sure about this.

**Comment by kenjis** [comment]
Level 9 is too high.

`phpstan.neon.dist`:
```
parameters:
	tmpDir: build/phpstan
	level: 5
	paths:
		- src/
	bootstrapFiles:
		- vendor/codeigniter4/framework/system/Test/bootstrap.php
	excludePaths:
		- src/Config/Routes.php
		- src/Views/*
	ignoreErrors:
	universalObjectCratesClasses:
		- CodeIgniter\Entity
		- CodeIgniter\Entity\Entity
		- Faker\Generator
	scanDirectories:
		- vendor/codeigniter4/framework/system/Helpers
		- vendor/codeigniter4/settings/src/Helpers
	dynamicConstantNames:
		- APP_NAMESPACE
		- CI_DEBUG
		- ENVIRONMENT
		- CodeIgniter\CodeIgniter::CI_VERSION
```
```console
$ XDEBUG_MODE=off phpstan analyse
...
 ------ ----------------------------------------------------------------------------------------------------------------------- 
  Line   Controllers/OAuthController.php                                                                                        
 ------ ----------------------------------------------------------------------------------------------------------------------- 
  22     Return type mixed of method Datamweb\ShieldOAuth\Controllers\OAuthController::redirectOAuth() is not covariant with    
         return type CodeIgniter\HTTP\RedirectResponse of method                                                                
         Datamweb\ShieldOAuth\Libraries\Basic\ControllersInterface::redirectOAuth().                                            
  48     Return type mixed of method Datamweb\ShieldOAuth\Controllers\OAuthController::callBack() is not covariant with return  
         type CodeIgniter\HTTP\RedirectResponse of method                                                                       
         Datamweb\ShieldOAuth\Libraries\Basic\ControllersInterface::callBack().                                                 
  84     Parameter #1 $name of function model expects class-string<CodeIgniter\Model>, string given.                            
  91     Call to an undefined method CodeIgniter\Model::findById().                                                             
  94     Call to an undefined method CodeIgniter\Model::addToDefaultGroup().                                                    
  124    Parameter #1 $name of function model expects class-string<CodeIgniter\Model>, string given.                            
  126    Call to an undefined method CodeIgniter\Model::findByCredentials().                                                    
  133    Parameter #1 $name of function model expects class-string<CodeIgniter\Model>, string given.                            
  134    Call to an undefined method CodeIgniter\Model::findByCredentials().                                                    
  147    Class Datamweb\ShieldOAuth\Controllers\LoginModel not found.                                                           
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                                                    
  147    Parameter #1 $name of function model expects class-string<CodeIgniter\Model>, string given.                            
  149    Call to an undefined method CodeIgniter\Model::recordLoginAttempt().                                                   
 ------ ----------------------------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------- 
  Line   Libraries/Basic/AbstractOAuth.php                                                            
 ------ --------------------------------------------------------------------------------------------- 
  28     Access to an undefined property Datamweb\ShieldOAuth\Libraries\Basic\AbstractOAuth::$token.  
  33     Access to an undefined property Datamweb\ShieldOAuth\Libraries\Basic\AbstractOAuth::$token.  
 ------ --------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------- 
  Line   Libraries/GithubOAuth.php                                                                    
 ------ --------------------------------------------------------------------------------------------- 
  29     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$client.         
  31     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$config.         
  32     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$callbake_url.   
  33     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$client_id.      
  34     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$client_secret.  
  39     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$callbake_url.   
  39     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$client_id.      
  46     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$client.         
  48     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$client_id.      
  49     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$client_secret.  
  51     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$callbake_url.   
  59     Caught class Datamweb\ShieldOAuth\Libraries\Exception not found.                             
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                          
  60     Call to method getMessage() on an unknown class Datamweb\ShieldOAuth\Libraries\Exception.    
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                          
  71     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$client.         
  79     Caught class Datamweb\ShieldOAuth\Libraries\Exception not found.                             
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                          
  80     Call to method getMessage() on an unknown class Datamweb\ShieldOAuth\Libraries\Exception.    
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                          
  90     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$config.         
  91     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$config.         
  92     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$config.         
  102    Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$config.         
  103    Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$config.         
  104    Access to an undefined property Datamweb\ShieldOAuth\Libraries\GithubOAuth::$config.         
  108    Variable $usersColumnsName might not be defined.                                             
 ------ --------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------- 
  Line   Libraries/GoogleOAuth.php                                                                  
 ------ ------------------------------------------------------------------------------------------- 
  32     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GoogleOAuth::$client.       
  34     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GoogleOAuth::$config.       
  49     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GoogleOAuth::$client.       
  62     Caught class Datamweb\ShieldOAuth\Libraries\Exception not found.                           
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                        
  63     Call to method getMessage() on an unknown class Datamweb\ShieldOAuth\Libraries\Exception.  
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                        
  74     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GoogleOAuth::$client.       
  82     Caught class Datamweb\ShieldOAuth\Libraries\Exception not found.                           
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                        
  83     Call to method getMessage() on an unknown class Datamweb\ShieldOAuth\Libraries\Exception.  
         💡 Learn more at https://phpstan.org/user-guide/discovering-symbols                        
  93     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GoogleOAuth::$config.       
  94     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GoogleOAuth::$config.       
  95     Access to an undefined property Datamweb\ShieldOAuth\Libraries\GoogleOAuth::$config.       
  106    Access to an undefined property Datamweb\ShieldOAuth\Libraries\GoogleOAuth::$config.       
  107    Access to an undefined property Datamweb\ShieldOAuth\Libraries\GoogleOAuth::$config.       
  108    Access to an undefined property Datamweb\ShieldOAuth\Libraries\GoogleOAuth::$config.       
  112    Variable $usersColumnsName might not be defined.                                           
 ------ ------------------------------------------------------------------------------------------- 

                                                                                                                        
 [ERROR] Found 52 errors  
```

**Pull request closed without merging**